framework/bundles/org.jboss.tools.rsp.secure/src/org/jboss/tools/rsp/secure/model/RSPSecureStorage.java

```diff
@@ -0,0 +1,169 @@
+package org.jboss.tools.rsp.secure.model;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.Arrays;
+import java.util.Properties;
+
+import org.jboss.tools.rsp.secure.crypto.CryptoException;
+import org.jboss.tools.rsp.secure.crypto.CryptoUtils;
+import org.jboss.tools.rsp.secure.crypto.NotInitializedCryptoException;
+
+public class RSPSecureStorage implements ISecureStorage {
+	private static final String COMMENT = "rsp secure storage";
+	private static final String HASH_COMMENT = "#" + COMMENT;
+	
+	
+	private File backingFile;
+	private byte[] key;
+	
+	private Properties secretData;
+	private CryptoUtils util;
+	
+	public RSPSecureStorage(File file, byte[] key) {
+		this.backingFile = file;
+		this.util = new CryptoUtils();
+		this.key = util.keyTo16(key);
+	}
+	
+	public void setKey(byte[] key) {
+		this.key = key;
+		secretData = null;
+	}
+	
+	public void load() throws CryptoException {
+		Properties tmp = new Properties();
+		if( backingFile != null && backingFile.exists()) {
+			try {
+				byte[] encrypted = util.getBytesFromFile(backingFile);
+				byte[] decrypted = util.decrypt(key, encrypted);
+				
+				byte[] magicBytes = HASH_COMMENT.getBytes(); 
+				if( !startsWith(decrypted, magicBytes)) {
+					throw new CryptoException("Invalid key", null);
+				}
+				tmp.load(new ByteArrayInputStream(decrypted));
+				Arrays.fill(decrypted, (byte)0);
+				secretData = tmp;
+			} catch(IOException | CryptoException e) {
+				throw new CryptoException("Unable to decrypt secure storage", e);
+			}
+		} else if( !backingFile.exists()) {
+			// new file, it's ok to create
+			secretData = tmp;
+		}
+	}
+	
+	private boolean startsWith(byte[] all, byte[] magic) {
+		for( int i = 0; i < magic.length; i++ ) {
+			if( i >= all.length) {
+				return false;
+			}
+			if( all[i] != magic[i]) {
+				return false;
+			}
+		}
+		return true;
+	}
+	
+
+	private boolean isInitialized() {
+		return secretData != null;
+	}
+	
+	public void save() throws CryptoException {
+		try {
+			if( backingFile != null && isInitialized()) {
```

Comment:
these 2 conditions would be easier to grasp if written like this:
```
if (!isInitialized) {
    throw new NotInitializedCryptoException("Secure storage not initialized", null);
}

if (backingFile != null) {
  ...
}
```

which then begs the question, what if it's initialized but has no backingfile?